### ISSUE TITLE
Design of MaterialOutputAction is probably incorrect

### ISSUE THREAD

**Issue by permcody** [idaholab/moose#12267]
## Rationale

Some of our developers have a need for creating custom syntax that includes automatically creating output objects (i.e. CSV) that are they themselves automatically populated with things like material property outputs. The current design of `MaterialOutputAction` inhibits this.

## Description

`MaterialOutputAction` currently retrieves a list of `add_output` `Actions` and dynamic casts them to an `AddOutputAction` type. This means that new `Actions` don't get picked up and can't work with this system in any way. 

I propose a solution where the `MaterialOutputAction` pulls the actually Output types and queries the instances directly. This avoids the extra indirection of pulling the Actions and also only have access to the InputParameters. New virtuals could be added to the Output class if necessary to provide the right user-extendable interface.

## Impact

Significant - The current design prevents developer extension.

Tag @aeslaughter, @sapitts

**Comment by permcody** [comment]
OK, let me clarify. It's not that we can't get something to work. The problem is conceptual. I'd rather correct the design before we start spreading more bad code. What you've already written should work (e.g. you are adding an output during the add_output task). Right now the MaterialOutputAction examines Actions to figure out what to do (e.g. what Variables and AuxKernels to add) instead of just looking at the objects themselves and making that decision. That's a really big problem! The extra level of indirection is unecessary.

Actually looking into this further I discovered another issue with it! It appears that we are adding AuxVariables in that Action as well, but that's long after the Action for adding AuxVariables has passed. This is unsupported and could easily break. This Action needs a serious overhaul. I'm sure we can get to this early next week.

Ugh... Setup debug is creating an Action instead of just creating an output. Why all the extra complication?

**Comment by permcody** [comment]
> The reason is that the variables that are going to be used for output need to be defined well before the Output objects are created, thus information is gathered from the actions, because the objects are not there to use.

Perhaps, except that you aren't doing that. You are adding the AuxVariables to the system in the "setup_material_output" task which is way late (long after the "add_aux_variable" task and after the system has been initialized. This wouldn't work at all if these were nonlinear variables. I'm actually surprised it works with the explicit variables. 

This Action needs to be associated with multiple tasks so it doesn't violate the initialization order. Since you discovered that aux variables can be indeed added very late, there might be a pretty good resolution to making this all work without exploring object params at all. Here's my proposed design:

Add a new "add_aux_output_variable" task. I know that moving the "add_aux_variable" task down the list won't work because there are several objects that query variables in the system to make decisions based on coupling and other needs. This new task will be allowed to add output variables only (those that can't or at least are not intended to be coupled to). This can replace the current "setup_material_output" task which is a little bit nebulous in terms of what it does. Additionally, we'll associate the current Action with "add_output" so that it'll add the outputs at the right time too. 

This will make it so that @sapitts new Action will be able to query outputs and get the list properly. I'll work on this.

**Comment by aeslaughter** [comment]
That might all be true and fine to do, but I talked with @sapitts, what you are proposing is not needed to help her. All she needs is an addOutput method on FEProblemBase. When I wrote the AddOutputAction I didn't want to add yet another function to FEProblem, but that is what we need. I can do this part. If you want to redesign the rest that is fine.

**Comment by permcody** [comment]
This is a step in the right direction. Now with that new method, we should be able to avoid creating meta-actions for adding outputs almost everywhere. Thanks.

**Comment by YaqiWang** [comment]
Notice that this could be an argument here. One think I cannot help to say is that *This has been in the code for years and has been working just fine* is not a good argument for not making changes... I think we all agreed that we strive for better designs in MOOSE to make the code both app developers and users happy.

**Comment by aeslaughter** [comment]
@YaqiWang I am not arguing that it shouldn't be changed. I didn't like it when I put it in 4 years ago, I am just tired of the constant negativity. We are all trying to write code that does a job, sometimes that code isn't perfect, but it works and gets the job done. As the framework develops and needs change the the code needs to change, fine, but we don't need to tear each other down.

I am as guilty as the next person for doing this, but it needs to stop, I will do my best to stop. @permcody just happened to hit a nerve with me on this, but what is done is done. This code needs to be changed to meet our needs, lets leave it at that.

**Comment by YaqiWang** [comment]
Yeah, understand, sometime I get frustrated and become emotional.
